src/Raven.Server/Documents/Indexes/MapReduce/AutoMapReduceIndex.cs

```diff
@@ -221,10 +210,20 @@ private unsafe void PutMappedResult(BlittableJsonReaderObject mappedResult, Lazy
             long storageId;
             if (existingEntry == null)
                 storageId = mapEntriesTable.Insert(tvb);
+            else if (existingEntry.ReduceKeyHash == reduceKeyHash)
+            {
+                // no need to update record in table since we have the same entry already stored
+                storageId = existingEntry.StorageId;
+            }
             else
             {
-                // no need to update since we have the same entry already stored
+                var previousState = GetReduceKeyState(existingEntry.ReduceKeyHash, indexContext, create: false);
+
                 storageId = existingEntry.StorageId;
+
+                previousState.Tree.Delete(new Slice((byte*)&storageId, sizeof(long)));
+
+                storageId = mapEntriesTable.Update(existingEntry.StorageId, tvb);
             }
 
             var pos = state.Tree.DirectAdd(new Slice((byte*)&storageId, sizeof(long)), mappedResult.Size);
```

Comment:
I think that now I understand what you mean. Instead of storing map results in leafs we can write them here to `mapEntriesTable` and the reducing executer will grab data from there by `storageId` stored as a key in the reduce tree. Correct?